### ISSUE TITLE
Segmentation fault when visiting

### ISSUE THREAD

**Issue by griseau** [vespa-engine/vespa#12068]
As mentioned on Gitter, I'm facing a segmentation fault when trying to use vespa-visiting. 

The command I'm executing is : `vespa-visit -v --priority LOW_1 -s "post.text_clean_with_stopwords=\"*skin*\"" --fieldset "post:permalink,text_raw,channel,uid" > /workspace/skin_documents.json`

In vespa.log only thing I can see is : 
`1580892627.908253	vespa-container	1072/6732	config-sentinel	config-sentinel.service	event	crash/1 name="searchnode" pid=3982 signal="Segmentation fault"`

Vespa version: 7.165.5 with only one node at the moment

**Comment by vekterli** [comment]
I can't see anything about the command itself that should indicate a potential problem.

Does this happen consistently if you run the command? A core dump backtrace of the offending thread would be very helpful to pinpoint the cause. If your system is not configured to collect core dumps, please consider temporarily enabling these (see the [Vespa documentation](https://docs.vespa.ai/documentation/reference/files-processes-and-ports.html#core-dumps) for an example of how to do so).

**Comment by griseau** [comment]
Yes tried it a few time, always having the same problem.

Not sure how to share it. Is there any way to retrieve the core dumps from a specified timestamp ?

**Comment by vekterli** [comment]
If you have configured `/proc/sys/kernel/core_pattern` the cores should be in the directory configured through it. Since this is very likely the same bug being hit every time, any core dump file should do.

If you have a file, say `proton-bi.core.1234`, something like this should give you the dump of the offending thread:
```
gdb --batch -ex bt $VESPA_HOME/sbin/vespa-proton-bin proton-bi.core.1234
```
Please let me know if this doesn't work

**Comment by griseau** [comment]
Looks ok, you can find the [dump.txt](https://github.com/vespa-engine/vespa/files/4159070/dump.txt). 

Let me know if you have any question

**Comment by vekterli** [comment]
Very interesting! Glob expressions are internally rewritten into regular expressions and it seems like something here has triggered the regex implementation into annihilating itself with a recursion so deep it runs out of stack space (touching a guard page and segfaulting the process). If I were to hazard a guess, this is likely due to the combination of document content and resulting regular expression triggering catastrophic backtracking during regex evaluation. Needless to say, this isn't optimal behavior.

I'll look more closely into this. For now, you could try working around the issue by using the explicit regex operator instead:
```
post.text_clean_with_stopwords=~"skin"
```
Since there are no anchors in the pattern, it should match if any substring is found.

**Comment by griseau** [comment]
Thanks for the workaround, it works this way.

**Comment by vekterli** [comment]
Looks like GCC's `<regex>` implementation is rather susceptible to this since it uses a recursive approach, re: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=86164

Since the "find a substring" glob case is so common, we could at least handle this more gracefully by implementing a rewrite rule at the document selection level.

Today we rewrite `*foo*` into `^.*foo.*$`, thereby sending GCC's regex engine screaming into a pit of doom.

Instead we could consider adding the following rewrite rules:
* `*foo*` -> `foo`
* `*foo` -> `foo$`
* `foo*` -> `^foo`

This isn't a generalized solution but avoids backtracking hazards for some of the most common matching patterns.

**Comment by griseau** [comment]
I just tried using `^.*foo.*$` and I face the same problem so not sure that could be a fix. 

Out of curiosity, where is it mapped in your codebase ?

**Comment by griseau** [comment]
My bad, I tried using `post.text_clean_with_stopwords=~\"^.*skin.*$\"` instead of post.text_clean_with_stopwords=\"^.*skin.*$\"`

So I confirm that `^.*foo.*$` could work

**Comment by vekterli** [comment]
The rewrite happens in https://github.com/vespa-engine/vespa/blob/master/document/src/vespa/document/select/operator.cpp#L195

It basically rewrites `*` -> `.*` and `?` -> `.`, escaping other possible regex control characters as it goes.

So the regex `^.*foo.*$` is exactly what you get when you write `*foo*` today, triggering the same behavior.

Note that the `=` operator is explicitly a glob operator. `==` is exact string match. `=~` is explicit regex operator. Using regex control characters with a `=` operator will cause such characters to be escaped, which is probably not what you want.

**Comment by vekterli** [comment]
@griseau resolving this took some more time than anticipated, but to ensure robustness we had to completely swap out the underlying regex engine in favor of one that is explicitly made to deal with all inputs in a well-defined manner (Google RE2). This should now be fixed at a core level from **7.188.5** and beyond.
